### PR TITLE
[CFU] Add summary entry point for FreeResponse contained levels

### PR DESCRIPTION
**Pull request by rshipp**
Add a button to the instructions panel in contained levels, where the contained level is a `FreeResponse` level type. The button links to the new summary view, added in #50606.

image

Supports RTL mode without modifications:

image

Known issue: it doesn't work very well at small screen sizes, or when zoomed in. Neither does the rest of the page.

image

## Links

* spec: [Summarizing Check for Understanding Levels](https://docs.google.com/document/d/1S3LYEBG9TwEcT2pwB-a9juV4topjd1polHIZ52i8aC8/edit#heading=h.u3h3ju25lkr)
* jira ticket: [TEACH-291](https://codedotorg.atlassian.net/browse/TEACH-291)

### CODE REVIEW

**Comment by dju90** [comment]
Looks great so far! Just curious - will this implementation work without modification for standalone free response levels as well?

**Comment by rshipp** [comment]
No, this is contained levels only. The `SummaryEntryPoint` component should work to be reused with minimal edits though.

File: apps/src/templates/levelSummary/SummaryEntryPoint.jsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  const data = getScriptData('freeresponse');
+
+  return (
+    <div className={styles.summaryEntryPoint}>
+      <Button
+        color="neutralDark"
+        text="View student responses"
+        icon="arrow-up-right-from-square"
+        href={summaryUrl}
+        __useDeprecatedTag
```
Comment (rshipp) [inline apps/src/templates/levelSummary/SummaryEntryPoint.jsx:27]: i'm using this flag because the `Button` styles push it down in the page for some reason:
image
but if it uses an `<a>` it looks normal. I'm not sure why `<a>` is considered a deprecated tag in this component; unlike `<div>`, this is actually an accessible implementation afaik.

File: dashboard/app/views/levels/_free_response.html.haml [hunk 2]
```diff
@@ -1,5 +1,12 @@
 :ruby
-  js_data = {}
+  js_data = {
+    responses: @responses&.map do |response|
+      {
+        user_id: response.user_id,
+        text: response.level_source.data,
```
Comment (rshipp) [inline dashboard/app/views/levels/_free_response.html.haml:6]: we don't use this yet, but it seemed easier to be consistent with the summary view. might want to take it out and just use `response_count` or something if we never end up using this data from the free response view. or, if anyone thinks I should take it out now and only add it back in if we need it, that's valid feedback too

File: apps/src/templates/levelSummary/CheckForUnderstanding.jsx [hunk 3]
```diff
@@ -46,7 +46,7 @@ const CheckForUnderstanding = ({
         {nextLevel && (
           <a
             className={isRtl ? styles.navLinkLeft : styles.navLinkRight}
-            href={`${nextLevel.url}${document.location.search}`}
+            href={nextLevel.url}
```
Comment (rshipp) [inline apps/src/templates/levelSummary/CheckForUnderstanding.jsx:43]: this and the `/summary` path changes are not technically related to this pr, but necessary to go out with it, since we don't want people navigating around to broken level summary pages

File: dashboard/app/controllers/script_levels_controller.rb [hunk 4]
```diff
@@ -145,6 +145,19 @@ def show
...
 
+    # TODO: If this adds too much to the load time in prod, move it to an API.
+    if current_user&.teacher?
+      @responses = []
+      # We use this for the level summary entry point, so on contained levels
+      # what we actually care about are responses to the contained level.
+      level = @level.contained_levels.any? ? @level.contained_levels.first : @level
+
+      # TODO: Change/remove this check as we add support for more level types.
+      if level.class.to_s == 'FreeResponse'
```
Comment (rshipp) [inline dashboard/app/controllers/script_levels_controller.rb]: i didn't see a const for this anywhere obvious - if anyone knows a cleaner way to check this i'd be happy to hear it

File: dashboard/config/routes.rb [hunk 5]
```diff
@@ -466,10 +466,16 @@
...
-            get 'summary', to: 'script_levels#show', defaults: {view: 'summary'}
           end
         end
+        resources :script_levels, only: [:show], path: "/levels", format: false do
+          # This route is defined in a separate resources, below the one above,
+          # because of how our assert_routing tests and Rails routing
+          # precedence work with multiple routes that point to the same action,
+          # with only a static path (no dynamic parts like 'path/:id').
+          # /s/xxx/lessons/yyy/levels/zzz/summary
+          get 'summary', on: :member, to: 'script_levels#show', as: 'summary', format: false, defaults: {view: 'summary'}
```
Comment (rshipp) [inline dashboard/config/routes.rb:477]: this took forever to figure out, and seems super hacky, since it's only to support the tests :/

File: dashboard/app/controllers/script_levels_controller.rb
Comment (rshipp) [inline dashboard/app/controllers/script_levels_controller.rb:1]: after doing some research, it seems to be a bad idea to have controller tests check instance variables, or routing configuration. if anyone has extra tests they'd like to see for these controller changes though, let me know

File: dashboard/app/controllers/script_levels_controller.rb [hunk 6]
```diff
@@ -131,7 +131,7 @@ def show
       return
     end
 
-    if request.path != (canonical_path = build_script_level_path(@script_level, @extra_params))
+    if request.path != (canonical_path = build_script_level_path(@script_level, @extra_params)) && params[:view] != 'summary'
```
Comment (rshipp) [inline dashboard/app/controllers/script_levels_controller.rb:134]: I checked if Maddie had any context on this, since she was the last to modify it - but as she pointed out, this code is actually unchanged from the [very first code commit](https://github.com/code-dot-org/code-dot-org/commit/7564066231e74aaeee763becfb3af70a84ce5dc5) in the repo. i did some manual testing with paths that trigger this (e.g. localhost-studio.code.org:3000/s/coursef-2020/lessons/2/levels/2/page/0), and it seems to be fine.

File: apps/src/templates/levelSummary/SummaryEntryPoint.jsx (hunk 1 above)
Comment (bethanyaconnor) [reply apps/src/templates/levelSummary/SummaryEntryPoint.jsx:27]: This is a known issue -- this tag was added to prevent using a `div` as a button but accidentally discouraged using `a` tags too. :( [receipt](https://github.com/code-dot-org/code-dot-org/pull/44815#discussion_r808387879)

File: dashboard/app/controllers/script_levels_controller.rb (hunk 4 above)
Comment (bethanyaconnor) [reply dashboard/app/controllers/script_levels_controller.rb]: Could you do something like `if level.is_a?(FreeResponse)`?

File: apps/src/templates/levelSummary/CheckForUnderstanding.jsx [hunk 7]
```diff
@@ -88,10 +83,8 @@ const CheckForUnderstanding = ({
           <p>
             <i className="fa fa-user" />
             <span>
-              {i18n.studentsSubmitted({
```
Comment (bethanyaconnor) [inline apps/src/templates/levelSummary/CheckForUnderstanding.jsx:91]: Is this string no longer needed? Can it be removed?

File: apps/src/templates/levelSummary/SummaryEntryPoint.jsx [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    `viewAs=${ViewType.Instructor}`
+  );
+  const summaryUrl = document.location.pathname + SUMMARY_PATH + params;
+
+  const data = getScriptData('freeresponse');
+
+  return (
+    <div className={styles.summaryEntryPoint}>
+      <Button
+        color="neutralDark"
```
Comment (bethanyaconnor) [inline apps/src/templates/levelSummary/SummaryEntryPoint.jsx]: nit: use `Button.ButtonColor.neutralDark`

File: apps/src/templates/levelSummary/SummaryEntryPoint.jsx [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        text="View student responses"
```
Comment (bethanyaconnor) [inline apps/src/templates/levelSummary/SummaryEntryPoint.jsx]: Should this be translated?

File: apps/src/templates/levelSummary/SummaryEntryPoint.jsx [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  students: PropTypes.arrayOf(
+    PropTypes.shape({
+      id: PropTypes.number.isRequired,
+      name: PropTypes.string
+    })
+  )
+};
+
+export default connect(
+  // NOTE: Some of this state data is loaded in by the teacher panel. If you
```
Comment (bethanyaconnor) [inline apps/src/templates/levelSummary/SummaryEntryPoint.jsx:54]: Thanks for this note!

File: apps/src/templates/levelSummary/summary-entry-point.module.scss [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  span.counter {
+    padding: 0 8px;
+    font-weight: bold;
+  }
+
+  span.text {
+    display: block;
+  }
+
+  /* stylelint-disable selector-pseudo-class-no-unknown */
```
Comment (bethanyaconnor) [inline apps/src/templates/levelSummary/summary-entry-point.module.scss:27]: just for my understanding: why is this needed?

File: apps/src/templates/levelSummary/SummaryEntryPoint.jsx [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+const SummaryEntryPoint = ({students}) => {
+  // If viewing the page as Participant, be sure to rewrite the link URL
+  // to view as Instructor, so we don't just get redirected back.
+  const params = document.location.search.replace(
+    `viewAs=${ViewType.Participant}`,
+    `viewAs=${ViewType.Instructor}`
+  );
+  const summaryUrl = document.location.pathname + SUMMARY_PATH + params;
+
+  const data = getScriptData('freeresponse');
```
Comment (bethanyaconnor) [inline apps/src/templates/levelSummary/SummaryEntryPoint.jsx]: it's interesting -- we normally get the script data in the entry point js, but I don't know if there's any reason to do that there vs here 🤔

File: dashboard/app/controllers/script_levels_controller.rb [hunk 13]
```diff
@@ -145,6 +145,19 @@ def show
     @level = select_level
     return if redirect_under_13_without_tos_teacher(@level)
 
+    # TODO: If this adds too much to the load time in prod, move it to an API.
+    if current_user&.teacher?
```
Comment (bethanyaconnor) [inline dashboard/app/controllers/script_levels_controller.rb:149]: We're moving towards a world where script levels pages are cached more frequently. That said, I don't believe we currently can support contained levels being cached and standalone FreeResponse/Multi levels aren't currently being cached to my knowledge. So, I think this is fine as is -- just wanted to think through this out loud.

File: apps/src/sites/studio/pages/levels/_free_response.js [hunk 14]
```diff
@@ -4,12 +4,30 @@ import React from 'react';
...
 import Attachments from '@cdo/apps/code-studio/components/Attachments';
 import {reportTeacherReviewingStudentNonLabLevel} from '@cdo/apps/lib/util/analyticsUtils';
+import {Provider} from 'react-redux';
+import InstructorsOnly from '@cdo/apps/code-studio/components/InstructorsOnly';
+import {getStore} from '@cdo/apps/redux';
 
 $(document).ready(() => {
   const data = getScriptData('freeresponse');
 
+  $('#containedLevel0 > #summaryEntryPoint').each(function() {
```
Comment (bethanyaconnor) [inline apps/src/sites/studio/pages/levels/_free_response.js:17]: To check my understanding -- this will only happen if there is a div with the id `#summaryEntryPoint` right? Which will only be there if the DCDO flag is `true`?

File: dashboard/app/views/levels/_free_response.html.haml [hunk 15]
```diff
@@ -30,6 +37,9 @@
       }
     );
 
+- if is_contained_level && DCDO.get("show_level_summary_entry_point", true)
```
Comment (bethanyaconnor) [inline dashboard/app/views/levels/_free_response.html.haml]: just checking -- do we want this DCDO flag to be true by default? ie is the plan to launch this when this PR hits prod and the DCDO is a switch in case we want to turn it off?

File: dashboard/app/controllers/script_levels_controller.rb (hunk 4 above)
Comment (rshipp) [reply dashboard/app/controllers/script_levels_controller.rb]: ahh, probably! good call, thanks

File: apps/src/templates/levelSummary/CheckForUnderstanding.jsx (hunk 7 above)
Comment (rshipp) [reply apps/src/templates/levelSummary/CheckForUnderstanding.jsx:91]: yeah, i wasn't sure if there were rules around when we remove things from the strings. it's unused though so i can take it out if that doesn't break anything

File: apps/src/templates/levelSummary/summary-entry-point.module.scss (hunk 11 above)
Comment (rshipp) [reply apps/src/templates/levelSummary/summary-entry-point.module.scss:27]: it's for the `:global(.fa)` selector, because the `fa` class doesn't come from inside this css module and would otherwise get a generated name that wouldn't match the element

File: apps/src/sites/studio/pages/levels/_free_response.js (hunk 14 above)
Comment (rshipp) [reply apps/src/sites/studio/pages/levels/_free_response.js:17]: yep! or if the flag is unset, because i defaulted it to true

File: dashboard/app/views/levels/_free_response.html.haml (hunk 15 above)
Comment (rshipp) [reply dashboard/app/views/levels/_free_response.html.haml]: that's my understanding, but @dju90 can you confirm?

Comment (dju90) [reply dashboard/app/views/levels/_free_response.html.haml]: Yes, but ideally I'd like to bug bash and launch the contained and standalone entry points simultaneously. Would that be possible?

Comment (rshipp) [reply dashboard/app/views/levels/_free_response.html.haml]: ah, in that case i think it'd be easier to default it to disabled, so I can merge this PR without waiting for the next one

Comment (bethanyaconnor) [reply dashboard/app/views/levels/_free_response.html.haml]: For the bug bash, is the plan to bash on staging or an adhoc before flipping the flag on production? (That makes sense as a plan to me, just want to check my understanding!)

File: apps/src/templates/levelSummary/summary-entry-point.module.scss (hunk 11 above)
Comment (bethanyaconnor) [reply apps/src/templates/levelSummary/summary-entry-point.module.scss:27]: gotcha thanks!

File: apps/src/templates/levelSummary/CheckForUnderstanding.jsx (hunk 7 above)
Comment (bethanyaconnor) [reply apps/src/templates/levelSummary/CheckForUnderstanding.jsx:91]: If it's definitely unused, it's okay to remove. For JS strings, the page will crash if the string isn't found, so we just want to be sure it's definitely not being used

File: apps/src/templates/levelSummary/SummaryEntryPoint.jsx (hunk 12 above)
Comment (rshipp) [reply apps/src/templates/levelSummary/SummaryEntryPoint.jsx]: i definitely copied this over from the entry point js when I moved the react code out of there. looking at the rest of the codebase, there are a ton of things in `apps/src/sites/studio` that are calling it, but only one in `apps/src/templates` outside of what I added for these new components. i'm not sure what the reason is, but i think that's enough of a pattern to justify me moving it back out of here and into the entry points

Comment (rshipp) [reply apps/src/templates/levelSummary/SummaryEntryPoint.jsx]: oh... as soon as i started implementing this i realized it makes the unit tests a million times easier since i can get rid of all the script-tag-injection stuff 😅 maybe that's why

File: dashboard/app/views/levels/_free_response.html.haml (hunk 2 above)
Comment (bethanyaconnor) [reply dashboard/app/views/levels/_free_response.html.haml:6]: How much harder would it be to use `response_count`? My only concern is that this is likely adding extra queries, but it's not so many that it's a big deal.

File: dashboard/config/routes.rb (hunk 5 above)
Comment (bethanyaconnor) [reply dashboard/config/routes.rb:477]: Thanks for figuring it out! And thanks for the comment :)

**Review by bethanyaconnor — APPROVED** [review]
🎉

File: dashboard/app/views/levels/_free_response.html.haml (hunk 2 above)
Comment (rshipp) [reply dashboard/app/views/levels/_free_response.html.haml:6]: it'd be a minor inconvenience to refactor this and the component. I can do it in the next pass i'm working on, if that's ok?

Comment (bethanyaconnor) [reply dashboard/app/views/levels/_free_response.html.haml:6]: sounds good! I'm also totally okay with a cleanup once this wraps up if we don't need it. You mentioned that it might be needed, so I'm happy with waiting until it's clear whether or not we will
